src/providers/tpm_provider/mod.rs

```diff
@@ -0,0 +1,462 @@
+        };
+        let key_data = serde_asn1_der::to_vec(&key).or_else(|err| {
+            error!("Could not serialise key elements: {}.", err);
+            Err(ResponseStatus::PsaErrorCommunicationFailure)
+        })?;
+
+        Ok(ResultExportPublicKey { key_data })
+    }
+
+    fn destroy_key(&self, app_name: ApplicationName, op: OpDestroyKey) -> Result<ResultDestroyKey> {
+        let key_name = op.key_name;
+        let key_triple = KeyTriple::new(app_name, ProviderID::TpmProvider, key_name);
+        let mut store_handle = self.key_id_store.write().expect("Key store lock poisoned");
+
+        let error_closure = |e| {
+            error!("Error storing a mapping: {}.", e);
+            Err(ResponseStatus::KeyIDManagerError)
+        };
+        if store_handle
+            .remove(&key_triple)
+            .or_else(error_closure)?
+            .is_none()
+        {
+            error!(
+                "Key triple \"{}\" does not exist in the Key ID Manager.",
+                key_triple
+            );
+            Err(ResponseStatus::KeyDoesNotExist)
+        } else {
+            Ok(ResultDestroyKey {})
+        }
+    }
+
+    fn asym_sign(&self, app_name: ApplicationName, op: OpAsymSign) -> Result<ResultAsymSign> {
+        let key_name = op.key_name;
+        let hash = op.hash;
+        let key_triple = KeyTriple::new(app_name, ProviderID::TpmProvider, key_name);
+
+        let store_handle = self.key_id_store.read().expect("Key store lock poisoned");
+        let mut esapi_context = self
+            .esapi_context
+            .lock()
+            .expect("ESAPI Context lock poisoned");
+
+        let len = hash.len();
+        if len > 64 {
+            error!("The buffer given to sign is too big. Its length is {} and maximum authorised is 64.", len);
+            return Err(ResponseStatus::PsaErrorInvalidArgument);
+        }
+
+        let password_context = get_password_context(&*store_handle, key_triple)?;
+
+        let signature = esapi_context
+            .sign(
+                password_context.context,
+                &password_context.auth_value,
+                &hash,
+            )
+            .or_else(|e| {
+                error!("Error signing: {}.", e);
+                Err(ResponseStatus::PsaErrorHardwareFailure)
+            })?;
+
+        Ok(ResultAsymSign {
+            signature: signature.signature,
+        })
+    }
+
+    fn asym_verify(&self, app_name: ApplicationName, op: OpAsymVerify) -> Result<ResultAsymVerify> {
+        let key_name = op.key_name;
+        let hash = op.hash;
+        let signature = op.signature;
+        let key_triple = KeyTriple::new(app_name, ProviderID::TpmProvider, key_name);
+
+        let store_handle = self.key_id_store.read().expect("Key store lock poisoned");
+        let mut esapi_context = self
+            .esapi_context
+            .lock()
+            .expect("ESAPI Context lock poisoned");
+
+        let len = hash.len();
+        if len > 64 {
```

Comment:
Ideally this should take into account the digest size of the hash algorithm that the key is created with - I think we actually have hash algorithms assigned to keys, maybe we can do something between the TPM crate and the provider to be able to get the public definition of the key and deduce from that the correct length

Reply:
Same for sign, of course

Reply:
Alternatively, these checks could be left to the crate to do by itself (which it actually is doing atm)

Reply:
Is not it fine at the moment as we are supporting only one kind of hash algorithm?

Reply:
Not sure about that! We actually support any hash algorithm, at least with Mbed Crypto; but it depends on what people want from us